lib/eclipse/EclipseState/Schedule/WellProductionProperties.cpp

```diff
@@ -42,67 +42,78 @@ namespace Opm {
+          the control mode specified in the keyword WCONHIST.
+
           There are a few ways to specify the BHPLimit in historical mode.
-          If the well is the under BHP control mode, then the observed BHP value will
+          If the well is under BHP control mode, then the observed BHP value will
           serve as the BHP limit.
 
-          If the control mode is not BHP, then by default the BHP limit is 1 atm. when declared
-          as history matching well for the first time with WCONHIST.
+          If the control mode is not BHP, then by default the BHP limit is 1 atm.
 
           You can reset the BHP limit to a specific value with keyword WELTARG after WCONHIST
           for both of the above situations.
 
           When the BHP limit is reset by WELTARG, the BHP limit will not be changed by the subsequent
           WCONHIST keyword if not under BHP control.
-
-          FBHPDEF can be used to set the default BHP limit, which is not implemented yet.
         */
 
-        const auto& cmodeItem = record.getItem("CMODE");
-        if (!cmodeItem.defaultApplied(0)) {
-            const std::string cmode_string = cmodeItem.getTrimmedString( 0 );
-            const auto cmode = WellProducer::ControlModeFromString(cmode_string);
+        namespace wp = WellProducer;
 
-            namespace wp = WellProducer;
+        // control mode will be used
+        wp::ControlModeEnum cmode = wp::CMODE_UNDEFINED;
 
-            if (cmode == wp::LRAT || cmode == wp::RESV || cmode == wp::ORAT ||
-                cmode == wp::WRAT || cmode == wp::GRAT || cmode == wp::BHP) {
-                p.addProductionControl( cmode );
-                p.controlMode = cmode;
-            } else {
-                const std::string well_name = record.getItem("WELL").getTrimmedString(0);
-                const std::string msg = "unsupported control mode " + cmode_string + " for WCONHIST of Well " + well_name;
-                throw std::invalid_argument(msg);
+        // when there is an effective control mode specified by WHISTCL, we always use this control mode
+        if ( effectiveHistoryProductionControl(controlModeWHISTCL) ) {
+            cmode = controlModeWHISTCL;
+        } else {
+            const auto& cmodeItem = record.getItem("CMODE");
+            if ( !cmodeItem.defaultApplied(0) ) {
+                const std::string cmode_string = cmodeItem.getTrimmedString( 0 );
+                cmode = wp::ControlModeFromString(cmode_string);
             }
+        }
 
-            // always have a BHP control/limit, while the limit value needs to be determined
-            p.addProductionControl( wp::BHP );
+        // through either WHISTCL or WCONHIST, we get a control mode
+        if ( effectiveHistoryProductionControl(cmode) ) {
+            p.addProductionControl( cmode );
+            p.controlMode = cmode;
+        } else {
+            const std::string cmode_string = wp::ControlMode2String(cmode);
```

Comment:
This line of code caused the test failure for the below record.  For a `SHUT` well, how much injection and production properties we should keep? Anyway, I will fix the failure here. 

```
 72 WCONHIST
 73    'W3'      'SHUT' /
 74 /
```